network/netplay/netplay_discovery.c

```diff
@@ -234,57 +255,102 @@ bool netplay_lan_ad_server(netplay_t *netplay)
       /* Somebody queried, so check that it's valid */
       addr_size = sizeof(their_addr);
 
-      if (recvfrom(lan_ad_server_fd, (char*)&ad_packet_buffer,
-            sizeof(struct ad_packet), 0, &their_addr, &addr_size) >=
-            (ssize_t) (2*sizeof(uint32_t)))
+      ret = recvfrom(lan_ad_server_fd, (char*)&ad_packet_buffer,
+            sizeof(struct ad_packet), 0, &their_addr, &addr_size);
+      if (ret >= (ssize_t) (2 * sizeof(uint32_t)))
       {
          char s[NETPLAY_HOST_STR_LEN];
          uint32_t content_crc         = 0;
 
          /* Make sure it's a valid query */
          if (memcmp((void *) &ad_packet_buffer, "RANQ", 4))
+         {
+            RARCH_LOG("[discovery] invalid query\n");
             continue;
+         }
 
          /* For this version */
          if (ntohl(ad_packet_buffer.protocol_version) !=
                NETPLAY_PROTOCOL_VERSION)
+         {
+            RARCH_LOG("[discovery] invalid protocol version\n");
             continue;
+         }
 
-         info              = runloop_get_system_info();
-
-         /* Now build our response */
-         content_crc = content_get_crc();
+         strlcpy(reply_addr, ad_packet_buffer.address, NETPLAY_HOST_STR_LEN);
 
-         memset(&ad_packet_buffer, 0, sizeof(struct ad_packet));
-         memcpy(&ad_packet_buffer, "RANS", 4);
+         net_ifinfo_t interfaces;
 
-         ad_packet_buffer.protocol_version =
-            htonl(NETPLAY_PROTOCOL_VERSION);
-         ad_packet_buffer.port = htonl(netplay->tcp_port);
-         strlcpy(ad_packet_buffer.retroarch_version, PACKAGE_VERSION,
-            NETPLAY_HOST_STR_LEN);
-         strlcpy(ad_packet_buffer.content, !string_is_empty(
-                  path_basename(path_get(RARCH_PATH_BASENAME))) 
-               ? path_basename(path_get(RARCH_PATH_BASENAME)) : "N/A",
-               NETPLAY_HOST_LONGSTR_LEN);
-         strlcpy(ad_packet_buffer.nick, netplay->nick, NETPLAY_HOST_STR_LEN);
+         if (!net_ifinfo_new(&interfaces))
+            return false;
 
-         if (info)
+         for (int k = 0; k < interfaces.size; k++)
          {
-            strlcpy(ad_packet_buffer.core, info->info.library_name,
-               NETPLAY_HOST_STR_LEN);
-            strlcpy(ad_packet_buffer.core_version, info->info.library_version,
-               NETPLAY_HOST_STR_LEN);
+            char *p;
+            char sub[NETPLAY_HOST_STR_LEN];
+
+            p=strrchr(reply_addr,'.');
+            if (p)
+            {
+               strlcpy(sub, reply_addr, p - reply_addr + 1);
+               if (strstr(interfaces.entries[k].host, sub) && !strstr(interfaces.entries[k].host, "127.0.0.1"))
```

Comment:
strstr will break on differing amounts of shorthand/zero padding with both v4 and v6 addresses, e.g. 127.000.000.001... better to match the integer value